File: Programmers/[77486] 다단계 칫솔 판매/JFe/77486.py
from collections import defaultdict
def solution(enroll, referral, seller, amount):
    answer = []
    link = defaultdict(list)    # 연결리스트(반대 방향)
    amount_dict = defaultdict(int)  # 판매원별 이익금 합계
    # 연결리스트 생성
    for i, r in enumerate(referral):
        if r == "-": continue
        link[enroll[i]].append(r)
    # 이익금 배분
    for i, a in enumerate(amount):
        cost = a * 100
        name = seller[i]
        while cost > 0:
            amount_dict[name] += cost - int(cost * 0.1)
            cost = int(cost * 0.1)
            if len(link[name]) == 0: break
            name = link[name][0]
    # answer 리스트에 추가
    for e in enroll:
        answer.append(amount_dict[e])
    return answer

File: Programmers/[77486] 다단계 칫솔 판매/SUbbb/77486.java
import java.util.*;

class Solution {
    private Map<String, Integer> sellerMoney = new HashMap<>();
    private Map<String, String> parent = new HashMap<>();
    private int len;

    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        len = enroll.length;
        initParent(enroll, referral);
        
        for (int i = 0; i < seller.length; i++)
            updateSellerMoney(seller[i], 100 * amount[i]);

        int[] answer = new int[len];
        for (int i = 0; i < len; i++)
            answer[i] = sellerMoney.getOrDefault(enroll[i], 0);
        
        return answer;
    }

    private void updateSellerMoney(String p, int money) {
        while(true) {
            if (p.equals("-") || money == 0) break;
            
            int dist = money / 10;
            sellerMoney.put(p, money - dist + sellerMoney.getOrDefault(p, 0));     
            p = parent.get(p);
            money = dist;
        }
    }

    private void initParent(String[] enroll, String[] referral) {
        for (int i = 0; i < len; i++)
            parent.put(enroll[i], referral[i]);
    }
}
